crates/ra_hir/src/ty/traits/chalk.rs

```diff
@@ -225,8 +255,29 @@ impl<'a, DB> chalk_solve::RustIrDatabase for ChalkContext<'a, DB>
 where
     DB: HirDatabase,
 {
-    fn associated_ty_data(&self, _ty: TypeId) -> Arc<AssociatedTyDatum> {
-        unimplemented!()
+    fn associated_ty_data(&self, id: TypeId) -> Arc<AssociatedTyDatum> {
+        debug!("associated_ty_data {:?}", id);
+        let type_alias: TypeAlias = from_chalk(self.db, id);
+        let trait_ = match type_alias.container(self.db) {
+            Some(crate::Container::Trait(t)) => t,
+            _ => panic!("associated type not in trait"),
+        };
+        let generic_params = type_alias.generic_params(self.db);
+        let parameter_kinds = generic_params
+            .params_including_parent()
+            .into_iter()
+            .map(|p| chalk_ir::ParameterKind::Ty(lalrpop_intern::intern(&p.name.to_string())))
```

Comment:
Heh, `lalrpop_intern::InternedString` is marked `Send`, which doesn't really makes sense because itnterner is thread local. 

I guess, in theory it could be useful, if you access interner on the one thread, and pass a reference to it to another thread, but that looks like a niche use-case.

OTOH, accidentally using the same index on two different threads seems like a footgun

Reply:
Hm, actually, could this be a real bug? We can call chalk from different threads, and, if it caches `InternedString` inside, that would lead to a wrong result!

Reply:
Yeah, good question... I was unhappy about having to add the lalrpop dependency anyway, ideally we wouldn't need to provide this at all. I expect Chalk doesn't actually use the name except for debug output...

@nikomatsakis @scalexm Do you know whether this is a problem? Would it be possible to remove the parameter names / make them optional?

Reply:
Ah, if that's only for debug it's centrally fine short-term. Long term, yeah, it would be cool to separate debug interface from the main interface.

Reply:
Hm, actually it turns out it does use the name: When there's an 'unselected' projection like `Vec::SomeAssociatedType` where it's not clear from which trait the associated type actually comes. So we aren't concerned with that yet, but we'll need to handle it somehow.

Maybe it'd be better to just have some `AssociatedTypeNameId` in Chalk and let us do the interning ourselves...